pt/lessons/advanced/gen-stage.md

```diff
@@ -0,0 +1,260 @@
+
+```elixir
+def start(_type, _args) do
+  import Supervisor.Spec, warn: false
+
+  children = [
+    worker(GenstageExample.Producer, [0]),
+    worker(GenstageExample.ProducerConsumer, []),
+    worker(GenstageExample.Consumer, []),
+  ]
+
+  opts = [strategy: :one_for_one, name: GenstageExample.Supervisor]
+  Supervisor.start_link(children, opts)
+end
+```
+
+Se tudo estiver certo, podemos executar nosso projeto e devemos ver tudo funcionando:
+
+```shell
+$ mix run --no-halt
+{#PID<0.109.0>, 2, :state_doesnt_matter}
+{#PID<0.109.0>, 4, :state_doesnt_matter}
+{#PID<0.109.0>, 6, :state_doesnt_matter}
+...
+{#PID<0.109.0>, 229062, :state_doesnt_matter}
+{#PID<0.109.0>, 229064, :state_doesnt_matter}
+{#PID<0.109.0>, 229066, :state_doesnt_matter}
+```
+
+Pronto! Como esperávamos, nossa aplicação apenas omite números pares e faz isso __rapidamente__.
+
+Neste ponto temos um pipeline funcionando. Existe um produtor emitindo números, um produtor-consumidor descartando números ímpares, e um consumidor mostrando tudo isso e continuando o fluxo.
+
+## Múltiplos Produtores ou Consumidores
+
+Mencionamos na introdução que era possível ter mais que um produtor ou consumidor. Vamos dar uma olhada nisso.
+
+Se examinarmos a saída do `IO.inspect/1` do nosso exemplo, vemos que todo evento é tratatdo por um único PID. Vamos fazer alguns ajustes para múltiplos _workers_ modificando o `lib/genstage_example/application.ex`:
+
+```elixir
+children = [
+  worker(GenstageExample.Producer, [0]),
+  worker(GenstageExample.ProducerConsumer, []),
+  worker(GenstageExample.Consumer, [], id: 1),
+  worker(GenstageExample.Consumer, [], id: 2),
+]
+```
+
+Agora que configuramos dois consumidores, vamos ver o que obtemos se rodarmos nossa aplicação agora:
+
+```shell
+$ mix run --no-halt
+{#PID<0.120.0>, 2, :state_doesnt_matter}
+{#PID<0.121.0>, 4, :state_doesnt_matter}
+{#PID<0.120.0>, 6, :state_doesnt_matter}
+{#PID<0.120.0>, 8, :state_doesnt_matter}
+...
+{#PID<0.120.0>, 86478, :state_doesnt_matter}
+{#PID<0.121.0>, 87338, :state_doesnt_matter}
+{#PID<0.120.0>, 86480, :state_doesnt_matter}
+{#PID<0.120.0>, 86482, :state_doesnt_matter}
+```
+
+Como você pode ver, agora nós temos múltiplos PIDs, simplesmente adicionando uma linha de código e dando IDs aos nossos consumidores.
+
+## Casos de Uso
+
+Agora que cobrimos o GenStage e construímos nossa primeira aplicação exemplo, quais são alguns casos de uso _reais_ do GenStage?
+
++ Pipeline de Transformação de Dados — Produtores não precisam ser simples geradores de números. Poderíamos produzir eventos de um banco de dados ou mesmo de outra fonte como Apache Kafka. Como uma combinação de produtor-consumidores e consumidores, podemos processar, ordenar, catalogar, e armazenar métricas à medida que elas ficam disponíveis.
+
++ Filas de trabalho — Uma vez que eventos podem ser qualquer coisa, poderíamos produzir unidades de trabalho para serem completadas por uma séries de consumidores.
```

Comment:
I would use title case here for "Filas de Trabalho" just as we used on the other items.